src/smart-auth/README.md

```diff
@@ -187,43 +193,62 @@ generateAuthorizationUri(
 The supported configuration for the provider config is just a TypeScript interface:
 
 ```typescript
-export type SmartAuthProvider = {
+export interface SmartAuthProvider {
   /** A name to label the provider */
   name: string;
-  /** @todo this could be typed to the FHIR spec */
-  scope: string[];
   /** Client registration */
   client: {
     id: string;
     secret: string;
-  }
-  /** Auth related config */
-  auth?: {
-    /** An optional prefix to add to every route path */
-    pathPrefix?: string;
-    /** Optional params to append to the authorization redirect */
-    authorizeParams?: Record<string, any>;
-    /** String used to set the host to request the tokens to. Required. */
-    tokenHost: string;
-    /** String path to request an access token. Default to /oauth/token. */
-    tokenPath?: string;
-    /** String path to revoke an access token. Default to /oauth/revoke. */
-    revokePath?: string;
-    /** String used to set the host to request an "authorization code". Default to the value set on auth.tokenHost. */
-    authorizeHost?: string;
-    /** String path to request an authorization code. Default to /oauth/authorize. */
-    authorizePath?: string;
   };
+  /** Auth related config */
+  auth: AuthCodeConfig | ClientCredentialsConfig;
+}
+```
+
+SMART Auth profiles can be configured for both the Authorization Code grant flow (i.e. as in CARIN BlueButton 2.0's SMART support) and, separately, for the Client Credentials grant flow introduced in some SMART + FHIR systems even though it is not standard.
+
+These are typed this way:
+
+```typescript
+interface SmartAuthConfig {
+  /** Supported grant flow */
+  grantFlow: GrantFlow;
+  /** String used to set the host to request the tokens to. Required. */
+  tokenHost: string;
+  /** String path to request an access token. Default to /oauth/token. */
+  tokenPath?: string;
+  /** Optional params to post to the token exchange */
+  tokenParams?: Record<string, any>;
+}
+
+interface ClientCredentialsConfig extends SmartAuthConfig {
+  grantFlow: "client_credentials"
+};
```

Comment:
for the `client_credientials` grant flow, are we banking on `tokenParams` having a scope property?

Reply:
No - if it has it, we will use it. But it doesn't need it.

In most OAuth 2.0 libraries, including simple-oauth2, you get to specify the "default" scope you want to use for a given Authorization Code grant flow request. When you want to support different grant flows with the same config, things get a little trickier because `scope` is _also_ used in the Credentials Grant flow.

How do they each work?

In the Authorization Code grant flow the `scope` value is used during the authorization URL creation step (the place we redirect the users to _first_). `scope` is commonly used as a global config value that is used for every single authorization URL, but there's no reason that you can't change it per request.

This might make sense for example when you have an app that flexibly lets the user select which scopes they want to give to the app. You could in advance ask the user what they want to give you, before redirecting them to the authorization URL.

The param `scope` _also_ happens to be used by the Client Credentials grant flow but it's very rarely ever going to make sense to have the same default values. Making a Client Credentials grant flow with scopes that are the same as a _user_ is sort of disjoint from the point of the CC grant flow: to give you application instead of user-level access to some resources. Worse, SMART App Launch as a spec says absolutely nothing about Client Credentials.

Reply:
So, if you set up a SmartAuthProvider with the Client Credentials grant flow, we're going to listen to anything you want to specify in `tokenParams` but no other defaults.